keps/sig-network/20190423-service-lb-finalizer.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Finalizer Protection for Service LoadBalancers
+authors:
+  - "@MrHohn"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloud-provider
+reviewers:
+  - "@andrewsykim"
+  - "@bowei"
+  - "@jhorwit2"
+  - "@jiatongw"
+approvers:
+  - "@andrewsykim"
+  - "@bowei"
+  - "@thockin"
+editor: TBD
+creation-date: 2019-04-23
+last-updated: 2019-04-23
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Finalizer Protection for Service LoadBalancers
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Release Signoff Checklist](#release-signoff-checklist)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+  - [Proposal](#proposal)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
```

Comment:
Can we add a section for "Risks and Mitigations"? I think we should discuss what happens if user does n+2 upgrade from v1.14 -> v1.16 and then does a downgrade back to v1.14. They would have added finalizers to the Service but then lose the removal logic on the downgrade. To be safe should we alpha for 2 releases or just assume a user would always do n+1 upgrade?

Reply:
Thanks for bringing this up. I was under the impression the supported upgrade order is always `1.n` to `1.(n+1)`? Ref https://kubernetes.io/docs/setup/version-skew-policy/#supported-component-upgrade-order.

Though should definitely cover that if `n+2` upgrade case is valid.

Reply:
> Thanks for bringing this up. I was under the impression the supported upgrade order is always 1.n to 1.(n+1)? Ref https://kubernetes.io/docs/setup/version-skew-policy/#supported-component-upgrade-order.

Yeah I believe officially it is 1 minor version that is supported but in practice/reality it's not uncommon for users to do n+2 :P. At the very least we should cover that in the KEP as a risk IMO even if we choose not to support it.

Reply:
SG, will add a "Risks and Mitigations" section for this.

Reply:
Added now.

Reply:
We should be careful to not try to support more than what is documented. n+2 is definitely not something that is intended to work.